Stricter ESLint configuration for code quality, consistency & cleaner merges

ybnd opened the pull request:
## References
* Enables a few ESLint rules that were disabled when porting from TSLint in https://github.com/DSpace/dspace-angular/pull/1567
* Fixes https://github.com/DSpace/dspace-angular/issues/583 (but for ESLint)
* Fixes https://github.com/DSpace/dspace-angular/issues/336



## Description
This PR makes our lint configuration a bit stricter, most importantly:

* [`@typescript-eslint/ban-types`](https://typescript-eslint.io/rules/ban-types/) → disallow confusing/harmful TypeScript types
* [`@angular-eslint/template/no-negated-async`](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin-template/docs/rules/no-negated-async.md) → disallow `!(value$ | async)` in Angular templates
* [`@angular-eslint/template/eqeqeq`](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin-template/docs/rules/eqeqeq.md) → disallow `!=` and `==` in Angular templates
* enforce [recommended RxJs rules](https://github.com/cartant/eslint-plugin-rxjs#rules) for consistency & safety
* [`indent`](https://eslint.org/docs/latest/rules/indent) & [`comma-dangle`](https://eslint.org/docs/latest/rules/comma-dangle) for reduced merge conflict noise
* import formatting for reduced merge conflict noise
  * alphabetical order (using [eslint-plugin-simple-import-sort](https://www.npmjs.com/package/eslint-plugin-simple-import-sort)  instead of [eslint-plugin-imports](https://www.npmjs.com/package/eslint-plugin-import) for better performance)
  * single line when importing a single object
  * multiline when importing multiple objects



## Instructions for Reviewers
Review whether the changes to the ESLint configuration are sensible

* If some of the rules are too strict/limiting, they can be changed or dropped
* Since this PR already forces _many_ formatting/code style changes, it's a good opportunity to make everything _more_ strict if we want to. Bundling everything in one PR would make it easier to incorporate.
  * e.g.: [Prettier](https://prettier.io/) would be a more consistent option for formatting, but would involve _even more_ conflicting changes. IMO it's worth it.
  * So far I've addressed the issues I've noticed, but there are bound to be more

Confirm whether the stricter rules are worth the increased linting time (from ~2 minutes to ~3 minutes)

The app (and the tests) should build and run in exactly the same way as before.



## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented:
@ybnd : wow!  Thanks for the effort here. 

Because this PR is so massive and touches literally everything, I recommend @artlowel or yourself bring this to an upcoming [DSpace Developer Meeting](https://wiki.lyrasis.org/display/DSPACE/Developer+Meetings) to discuss.  It seems like something we need to come to agreement on quickly, otherwise this PR will just continually have merge conflicts when any code changes.

Let me know if there's a particular meeting where we should add this to the agenda.

atarix83 reviewed (commented):
thanks @ybnd 

I reviewed the new rules configured for and i think they are overall good and we'll help us to maintain the code cleaner.

My only doubt is on the  `@angular-eslint/template/no-negated-async`, indeed the condition 
`!(value$ | async)` is asserted when the observable value is undefined, null or false. 
Looking at the changes you have done not always is possible to be sure which one was the intentional assertion, especially between null and undefined. So at least we should add both condition  `(value$ | async) === undefined || (value$ | async) === null` but honestly I prefer the shorter way `!(value$ | async)`.

I'd hear other opinion on this rule, but honestly i'd avoid to use it.

tdonohue commented:
@ybnd : as of today all major features for 8.0 are merged.  So, it'd be good to get this PR updated as soon as reasonably possible if we want to merge it before other refactoring/upgrade work.

@artlowel : In order to ensure this can be merged in a timely manner, I'd appreciate it if you could provide your feedback into this PR, especially the discussion between @atarix83 and @ybnd about whether to include `@angular-eslint/template/no-negated-async` above. 

My goal is to get this merged as soon as possible, but that will require coming to a quick agreement on what should be "in scope" vs "out of scope" for these stricter ESLint settings.  Please do keep in mind that we can also change these settings again in the future if needed (e.g. if we make them too strict or not strict enough), so the key point is to come to a **quick consensus**.

artlowel commented:
I'd prefer to keep the `@angular-eslint/template/no-negated-async` rule. We also discourage the use of `!variable` in regular typescript code in favor of explicit checks for `null`, `undefined` and `false`, because that makes it explicit what is meant, and prevents bugs. It makes sense to do the same thing for templates in my opinion.

And as @ybnd said and [the documentation for the rule](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin-template/docs/rules/no-negated-async.md) explains: it can prevent performance issues and flashing content, because async pipes start off by emitting `null`

tdonohue commented:
Because of the potential **benefits** that @ybnd and @artlowel have pointed out in the `@angular-eslint/template/no-negated-async` rule [documentation](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin-template/docs/rules/no-negated-async.md), I agree that I'd like to start with this rule **enabled**.   If, however, we find it becomes a problem or hassle for developers, we can always disable it at a later date.   Hopefully that seems reasonable to @atarix83 as well.

@ybnd : Could you please update this PR as soon as possible?  Ideally, I'd like to get this merged immediately in order to get #2750 freed up to also be merged quickly.  Thanks.

atarix83 reviewed (dismissed):
@ybnd 

If everyone agreed we should keep the know rule. Please just check the cases reported in my inline comments. Thaks

tdonohue commented on src/app/collection-page/edit-collection-page/collection-source/collection-source.component.html:
```diff
@@ -1,7 +1,7 @@
-        <button class=" btn btn-danger" *ngIf="!(isReinstatable() | async)"
-                [disabled]="!(hasChanges() | async)"
+        <button class=" btn btn-danger" *ngIf="(isReinstatable()  | async) === false"
+                [disabled]="(hasChanges()  | async) === false"
```
@ybnd & @artlowel  : I think @atarix83 is correct here that this may not work properly if we want to keep this rule enabled... see also this [stackoverflow answer](https://stackoverflow.com/questions/66789429/eslint-async-pipes-should-not-be-negated/67971535#67971535).   In that Stackoverflow answer, they also recommend using either `!== true` or possibly creating a custom pipe to handle this better (`| isTruthy`).   

I think we need to come to a **quick solution** to these questions or consider temporarily removing this rule and add it back in once we have a solution.  I don't want to delay the work on #2750 or upgrading to Angular 16 for much longer...so if we don't come to a solution, this work may require delaying.

(Apologies for rushing this, but we are very behind schedule for these upgrades & I'll be out-of-office on March 13-17, so I feel we must move this forward quickly)

ybnd replied:
@tdonohue @atarix83 `!== true` makes sense
I've addressed most of the comments, will go through them once more tomorrow and push before (CET) noon

ybnd commented:
@tdonohue @atarix83 @artlowel 
I've addressed the comments above
- Replaced `(... | async) === false` cases with `(... | async) !== true`
  - Haven't found any obvious situations where this would cause problems, so I think it's easier if it's consistent
  - There are a few situations where the source observable is a `BehaviorSubject<boolean>` that never emits null or undefined, so the change is not even necessary, but again: consistency
- Added two new pipes in case we want to check null/undefined for arbitrary types: `dsHasValue` and `dsHasNoValue`
  - Less of a hassle to write out, and we use the equivalent `hasValue`/`hasNoValue` methods all over the place already, so this feels like the way to go

ybnd commented on src/app/access-control/epeople-registry/eperson-form/eperson-form.component.html:
```diff
@@ -47,13 +47,13 @@ <h1 class="border-bottom pb-2">{{messagePrefix + '.edit' | translate}}</h1>
         <h2>{{messagePrefix + '.groupsEPersonIsMemberOf' | translate}}</h2>
 
-        <ds-themed-loading [showMessage]="false" *ngIf="!(groups | async)"></ds-themed-loading>
+        <ds-themed-loading [showMessage]="false" *ngIf="(groups$ | async) === undefined"></ds-themed-loading>
```
Replaced with `| hasNoValue` (same for a few other equivalent cases)

ybnd commented:
(force pushed last two commits because I'd committed a temporary Karma config change by accident)

tdonohue reviewed (commented):
@ybnd: I'm reviewing/testing this today.  Overall, it looks good, and I'm not seeing any changes in behaviour when running the UI in production mode (`yarn build:prod; yarn serve:ssr`).  **However**, I am seeing that all HTTP errors (404s, etc) are now logged in **great detail** in the SSR logs (in console where you ran `yarn serve:ssr`) and in the browser DevTools Console.  Is there maybe a stray `console.log` in this PR (I'm having difficulty finding it myself)?  Somehow this PR makes those logs a lot more noisy.

Beyond that, I haven't found any other users, but I'm still testing various areas of the application.

tdonohue commented:
@ybnd : Regarding my [prior comment](https://github.com/DSpace/dspace-angular/pull/2343/#pullrequestreview-1923039857), I think the stray `console.log` is in `dspace-rest.service.ts` on line 113 (in the "catchError" block).  It should likely be removed as it's very **noisy**.  It results in every HTTP 404/400 to be logged, which I don't think we should do.

Once that log is removed, I'm +1 this PR.  I haven't found any obvious bugs or changes in behavior in my testing today.

ybnd commented:
@tdonohue thanks for pointing that out, looks like I left it there after testing
I'll take a quick look around to ensure there are no other new `console.log`s that are less noisy maybe

tdonohue reviewed (approved):
👍 Thanks @ybnd!  This looks good to me now. (I also agree that restoring the behavior of that console.log seems good enough for now & we can set that aside for a different PR/ticket)

tdonohue commented:
Merging this immediately in order to free up work on #2750 (as discussed in yesterday's DevMtg).

@ybnd : Now that this is merged, can you please help to enhance/update the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide) (Angular section) with details about the new style rules from this PR?   At a minimum, I think we should document the following:
* Don't use `!([variable] | async)` or `([variable] | async) === false`.  Instead use `([variable] | async) !== true`
* Similar instructions on when to appropriately use `dsHasValue` and `dsHasNoValue` pipes
* Other major code rules (e.g. ordering of imports) you feel should be documented for developers.  (We don't need to document every single thing that lint enforces, but it would be nice to add some basic notes.)

If you don't feel comfortable updating this wiki page, you can add notes in a comment here & I'll copy them over.  In the meantime, I'll flag this as `needs documentation`.
